### PR TITLE
[Merged by Bors] - Sync committee sign bn fallback

### LINKED ISSUE #3612: VC does not fallback for sync sigs when primary BN is optimistic
## Description

When the VC attempts to publish a sync committee signature it will try and get the head block root from the first BN that returns a response, regardless if that BN is optimistic or not:

https://github.com/sigp/lighthouse/blob/01e84b71f524968f5b940fbd2fa31d29408b6581/validator_client/src/sync_committee_service.rs#L177-L209

I think the ideal behaviour would be trying to find a BN that has a *non-optimistic* head block root. At first glance I think this would be pretty easy to achieve by pushing the `execution_optimistic` checks inside the `first_success` function.

### PR DESCRIPTION
**Pull request by tthebst**
## Issue Addressed

Closes #3612

## Proposed Changes

- Iterates through BNs until it finds a non-optimistic head.

A slight change in error behavior: 
- Previously: `spawn_contribution_tasks` did not return an error for a non-optimistic block head. It returned `Ok(())` logged a warning.
- Now: `spawn_contribution_tasks` returns an error if it cannot find a non-optimistic block head. The caller of `spawn_contribution_tasks` then logs the error as a critical error.

### CODE REVIEW

File: validator_client/src/sync_committee_service.rs [hunk 1]
```diff
@@ -175,26 +175,28 @@ impl<T: SlotClock + 'static, E: EthSpec> SyncCommitteeService<T, E> {
...
+            .first_success(
+                RequireSynced::Yes,
+                OfflineOnFailure::Yes,
+                |beacon_node| async move {
+                    match beacon_node.get_beacon_blocks_root(BlockId::Head).await {
+                        Ok(Some(block)) if block.execution_optimistic == Some(true) => {
+                            Ok(block.data.root)
+                        }
+                        Ok(Some(_)) => {
+                            Err(format!("To sign sync commitee messages for slot {} a non-optimistic head block is required",slot))
```
Comment (michaelsproul) [inline validator_client/src/sync_committee_service.rs]:
```suggestion
                            Err(format!("To sign sync committee messages for slot {slot} a non-optimistic head block is required"))
```

**Review by michaelsproul — APPROVED** [review]
Nice!

Sorry it took so long for us to review, will merge this with the typo corrected.

**Comment by michaelsproul** [comment]
noticed that the simulator is failing semi-consistently with errors about sync aggregates. it may not be related to this PR but I think we should look into it before merging

File: validator_client/src/sync_committee_service.rs [hunk 2]
```diff
@@ -174,39 +174,40 @@ impl<T: SlotClock + 'static, E: EthSpec> SyncCommitteeService<T, E> {
...
-        let block_root = response.data.root;
-        if let Some(execution_optimistic) = response.execution_optimistic {
-            if execution_optimistic {
+        let block_root = match response {
+            Ok(block) => block.data.root,
+            Err(errs) => {
                 warn!(
                     log,
-                    "Refusing to sign sync committee messages for optimistic head block";
+                    "Refusing to sign sync committee messages for an optimstic head block or \
```
Comment (michaelsproul) [inline validator_client/src/sync_committee_service.rs]:
```suggestion
                    "Refusing to sign sync committee messages for an optimistic head block or \
```

**Review by michaelsproul — APPROVED** [review]
Looks good to go now! Good catch on the flipped condition, sorry I missed that!

**Comment by tthebst** [comment]
The if statement below that checked if `optimistic_execution==None` was also an issue. This case is now an error in and not returned in the `first_sucess` iterator.

**Comment by michaelsproul** [comment]
It'll be a breaking change for old versions of Lighthouse that don't have that flag, but should only affect Gnosis chain (which is still unmerged).

**Pull request closed without merging**